curriculum/challenges/english/20-upcoming-python/learn-tree-traversal-by-building-a-binary-search-tree/655a15b683445611528cccf1.md

```diff
@@ -0,0 +1,48 @@
+---
+id: 655a15b683445611528cccf1
+title: Step 18
+challengeType: 20
+dashedName: step-18
+---
+
+# --description--
+
+At the end of your `_insert` method, after the insertion process is complete, return the current node to update the tree structure at the higher levels of the recursive call stack.
+
+
+# --hints--
+
+You should return the current node outside the conditional blocks.
+
+```js
+({ test: () => assert.match(e.code.original["main.py"], /^\s{8}return\s+node/m) })
+```
+
+# --seed--
+
+## --seed-contents--
+
+```py
+class TreeNode:
+    def __init__(self, key):
+        self.key = key
+        self.left = None
+        self.right = None
+
+class BinarySearchTree:
+    def __init__(self):
+        self.root = None
+
+    def insert(self,key):
+        self.root = self._insert(self.root, key)
+
+--fcc-editable-region--
+    def _insert(self, node, key):
+        if node is None:
+            return TreeNode(key)
+        if key < node.key:
+            node.left = self._insert(node.left, key)
+        elif key > node.key:
+            node.right = self._insert(node.right, key)                
```

Comment:
```suggestion
        elif key >= node.key:
            node.right = self._insert(node.right, key)                
```
Should the code also handle cases when the `key` is equal to `node.key`? If the tree may have duplicates, this condition could be updated to use the `>=` operator.

Reply:
Yes, there isn't a separate sace for equality. But, the code implicitly handles the case of a duplicate key by not performing a recursive call in that situation and returns the current node unchanged.